hal/network/lwip/wiznet/wiznetif_config.cpp

```diff
@@ -0,0 +1,163 @@
+
+#if HAL_PLATFORM_NCP
+#include "system_cache.h"
+#endif // HAL_PLATFORM_NCP
+
+namespace particle {
+
+namespace net {
+
+using namespace particle::services;
+
+WizNetifConfig::WizNetifConfig() :
+        initialized_(false) {
+}
+
+void WizNetifConfig::init() {
+    memset(&wizNetifConfigData_, 0, sizeof(wizNetifConfigData_));
+    recallWizNetifConfigData();
+    if (validateWizNetifConfigData() != SYSTEM_ERROR_NONE) {
+        recallWizNetifConfigData();
+    }
+    initialized_ = true;
+}
+
+// void WizNetifConfig::logWizNetifConfigData(WizNetifConfigData& configData) {
+//     LOG(INFO, "wizNetifConfigData size:%u cs_pin:%u reset_pin:%u int_pin:%u",
+//             configData.size,
+//             configData.cs_pin,
+//             configData.reset_pin,
+//             configData.int_pin);
+// }
+
+int WizNetifConfig::saveWizNetifConfigData() {
+    WizNetifConfigData tempData = {};
+    int result = SystemCache::instance().get(SystemCacheKey::WIZNET_CONFIG_DATA, (uint8_t*)&tempData, sizeof(tempData));
+    if (result != sizeof(tempData) ||
+            /* memcmp(tempData, wizNetifConfigData_, sizeof(tempData) != 0)) { // not reliable */
+            tempData.size != wizNetifConfigData_.size ||
+            tempData.cs_pin != wizNetifConfigData_.cs_pin ||
+            tempData.reset_pin != wizNetifConfigData_.reset_pin ||
+            tempData.int_pin != wizNetifConfigData_.int_pin ) {
+        // LOG(INFO, "Saving cached wizNetifConfigData, size: %d", wizNetifConfigData_.size);
+        // logWizNetifConfigData(wizNetifConfigData_);
+        result = SystemCache::instance().set(SystemCacheKey::WIZNET_CONFIG_DATA, (uint8_t*)&wizNetifConfigData_, sizeof(wizNetifConfigData_));
+    }
+    return (result < 0) ? result : SYSTEM_ERROR_NONE;
+}
+
+int WizNetifConfig::recallWizNetifConfigData() {
+    WizNetifConfigData tempData;
+    int result = SystemCache::instance().get(SystemCacheKey::WIZNET_CONFIG_DATA, (uint8_t*)&tempData, sizeof(tempData));
+    if (result != sizeof(tempData)) {
+        return SYSTEM_ERROR_NOT_FOUND;
+    }
+    // LOG(INFO, "Recalling cached wizNetifConfigData");
+    // logWizNetifConfigData(tempData);
+    memcpy(&wizNetifConfigData_, &tempData, sizeof(wizNetifConfigData_));
+    return SYSTEM_ERROR_NONE;
+}
+
+// int WizNetifConfig::deleteWizNetifConfigData() {
+//     LOG(INFO, "Deleting cached wizNetifConfigData");
+//     int result = SystemCache::instance().del(SystemCacheKey::WIZNET_CONFIG_DATA);
+//     return (result < 0) ? result : SYSTEM_ERROR_NONE;
+// }
+
+int WizNetifConfig::setConfigData(const WizNetifConfigData* configData) {
+    WizNetifConfigLock lk;
+
+    if (!initialized_) {
+        init();
+    } else {
+        validateWizNetifConfigData();
+    }
+
+    if (configData) {
+        memcpy(&wizNetifConfigData_, configData, sizeof(WizNetifConfigData));
+        // logWizNetifConfigData(wizNetifConfigData_);
+        saveWizNetifConfigData();
+    } else {
+        return SYSTEM_ERROR_BAD_DATA;
```

Comment:
Prefer `SYSTEM_ERROR_INVALID_ARGUMENT`

Reply:
If we add a check for pin value, can we still bundle these two errors into BAD DATA or is INVALID ARGUMENT better, or separate errors for each error type?

Reply:
I'd prefer returning `SYSTEM_ERROR_NONE` and restore the default pin values.

Reply:
We should return SYSTEM_ERROR_INVALID_ARGUMENT if the user tries to input invalid data, but we can also restore defaults as well in that case.